pkg/machine/define/vmfile.go

```diff
@@ -50,14 +49,15 @@ func (m *VMFile) Read() ([]byte, error) {
 }
 
 // Read the first n bytes of a given file and return in []bytes
-func (m *VMFile) Peek(n int) ([]byte, error) {
+func (m *VMFile) ReadMagicNumber(n int) ([]byte, error) {
 	f, err := os.Open(m.GetPath())
 	if err != nil {
 		return nil, err
 	}
 	defer f.Close()
-	r := bufio.NewReader(f)
-	return r.Peek(n)
+	b := make([]byte, n)
+	_, err = f.Read(b)
+	return b, err
```

Comment:
This doesn’t correctly handle files shorter than `n` bytes (and may return `io.EOF` with files exactly `n` bytes, which the caller then doesn’t handle)

Reply:
```go
n, err = f.Read(b)
if err != nil && err != io.EOF { /* fail */ }
return b[:n], nil
```
I think

Reply:
… something using `io.ReadAtLeast`, actually? I can’t see any Go guarantee that `Read` will not cause short reads.

Reply:
`io.ReadFull`, even.

It’s fascinating(?) how many iterations I need to get this right.

Reply:
but would ReadFull even be correct? It may work in practice but consider we have some new compression arg with 100 bytes magic, and then another compressed file that is less than 100 bytes than you would always error with ReadFull which seem incorrect.

Reply:
ah i see io.ReadFull will return ErrUnexpectedEOF in this case so that should then be ignored.

Reply:
Yes, `io.ReadFull` + ignore `EOF` + ignore `ErrUnexpectedEOF` (but fail on other errors).

Reply:
This makes sense in general but I am wondering: wouldn't it be better to fail right away if the machine VM file is shorter than or equal to `magicNumberMaxSize`?

Reply:
Given that the function is now named specifically `ReadMagicNumber`, sure, that works.

But a `ReadFull`, or an equivalent loop, still seems to be required — or, quite possibly, I’m missing some place where Go promises not to return short reads.

Reply:
From stdlib `(os.File).Read(b)` doc:

> Reads up to len(b) bytes from the File and stores them in b.

==> if the file is shorter than n, it still reads it content without returning any error.

> At end of file, Read returns 0, io.EOF.

==> io.EOF is returned only when there is nothing else to read.

Now, I can still handle `io.EOF` specifically when it's returned by `ReadMagicNumber` but I would not ignore it.

Reply:
- “up to `len(b)`” would be satisfied by returning just 1 byte, not enough for our magic detection.
- It’s _not_ guaranteed that `io.EOF` is returned _only_ with `n == 0`. https://pkg.go.dev/io#Reader explicitly allows the other behavior, and I can’t see anything for `os.File.Read` documented to override that.